### ISSUE TITLE
Golang Client

### ISSUE THREAD

**Issue by nightrune** [OpenLightingProject/ola#728]
Basic Go Client and Build environment

**Comment by apinnecke** [comment]
Hey @nightrune , have you done any work here? Otherwise I'd chime in and build a Golang client since I actually need one right now. Any docs or notes you've taken that you could provide to make the job more easy? thanks in advance :)

**Comment by nightrune** [comment]
Oh man! I got pretty far. The hard part was writing the protobuf compiler.
I'm not sure where I left that. You should be able to see where I left off
on my fork. It's quite a bit behind from master now though.

https://github.com/nightrune/ola/tree/go_client?files=1

I'd love to see it done though.

Also to get the client to build with make and the rest of the project there
is a script that changes and moves $GOCODE.

Let me know if you have anymore questions. We had decided that it would be
a blocking API, and I was writing it as I learned Go. It may not be the
best design.

**Comment by apinnecke** [comment]
Wait, what? You've written a protobuf compiler? 😮  May I asked why you didn't use the official one?

Do you have any time to continue the work? I've cloned your fork and will have a look on your go code to see in which state it is.

**Comment by nightrune** [comment]
Haha yeah! It uses the official one for the raw protobuf packets but ola's
c++ client has a custom compiler extension for the RPC layer, and that's
what I built/was building.

I've got very very little to build on it. I'm a contractor and business has
been bad (probably my fault) so the amount of free time is not there. Happy
to guide anyone else through though. Just don't have the time to put my
head deep into code.

On Mar 11, 2017 8:40 AM, "Alexander Pinnecke" <notifications@github.com>
wrote:

Wait, what? You've written a protobuf compiler? 😮 May I asked why you
didn't use the official one?

Do you have any time to continue the work? I've cloned your fork and will
have a look on your go code to see in which state it is.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/OpenLightingProject/ola/issues/728#issuecomment-285879627>,
or mute the thread

**Comment by apinnecke** [comment]
> ... and I was writing it as I learned Go. It may not be the best design.
Oh ... yah, I see. TBH, that code is far from Golang code style, golint and gofmt are completely raging. :/

Well, I'd love to chime in here ... the issue is, that I am a bloody new guy on the OLA project and could need a crash course on conventions and those gotchas, but I know Golang very well. So ... how about comparing power here man?

**Comment by nightrune** [comment]
Yeah at the time it seemed better to follow OLA style then go style. I'd
likely not do that now. I'm glad to help as much as possible with OLA
internals as I can. We can get @nomis52 and Peter in on it as well.

I'm pretty busy this week. Let me make time next Saturday (hopefully
sooner) to write out what I know.

The short is, The actual protobuf packets are pretty simple. The RPC API
tends to be 1to1 with the packets. If you know c++ and python well you can
look at those clients. Very little fancy logic happens in the client code.
Most of what the client is, is hiding protobuf and making it into function
calls.

If I remember right, the plan for the golang client was to have an
ola.Client struct. When you Run() it it fires off a read and write go
routine that each have a channel. So on a new packet from the server you
might do basic checks on the packet but it would just send it to the client
through the channel. Sending is pretty easy as you pack and setup with
protobuf the send to the TX channel.

After that layer is done we can look at adding support for the settings
side on plugins. Artnet has a set of sub commands that you can send as well.

On Mar 11, 2017 9:02 AM, "Alexander Pinnecke" <notifications@github.com>
wrote:

... and I was writing it as I learned Go. It may not be the best design.
Oh ... yah, I see. TBH, that code is far from Golang code style, golint and
gofmt are completely raging. :/

Well, I'd love to chime in here ... the issue is, that I am a bloody new
guy on the OLA project and could need a crash course on conventions and
those gotchas, but I know Golang very well. So ... how about comparing
power here man?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/OpenLightingProject/ola/issues/728#issuecomment-285880998>,
or mute the thread

**Comment by apinnecke** [comment]
Okay, great, looking forward to have that - until then I could just use the JSON api to do the same, right?

**Comment by nightrune** [comment]
Yeah the HTTP/JSON layer should be pretty complete. Peter would know more
the me. As he maintains it right now.

**Comment by apinnecke** [comment]
Great, will use that one for the moment. Thanks for all the info and all the work and time, man! 💪

**Comment by peternewman** [comment]
Yes the JSON stuff should do most of what you want @apinnecke , depending on what you want to do. It's probably not the fastest or best though.

If you do a protobuf client, be aware of https://github.com/OpenLightingProject/ola/issues/1192 . You'll either be forced to use the latest version of protobuf, or stuck on the old one.

See the gory details/docs for the RPC system here:
https://docs.openlighting.org/ola/doc/latest/rpc_system.html
